test/service/src/main.rs

```diff
@@ -0,0 +1,192 @@
+use std::{io::Write, sync::Arc};
+
+fn extract_genesis_wasm(chain_spec: &Box<dyn sc_service::ChainSpec>) -> Result<Vec<u8>, String> {
+	let mut storage = chain_spec.build_storage()?;
+
+	storage
+		.top
+		.remove(sp_core::storage::well_known_keys::CODE)
+		.ok_or_else(|| "Could not find wasm file in genesis state!".into())
+}
+
+pub fn wrap_announce_block() -> Box<dyn FnOnce(AnnounceBlockFn) -> AnnounceBlockFn> {
+	Box::new(|_| {
+		// Never announce any block
+		Arc::new(|_, _| {})
+	})
+}
+
+fn main() -> Result<(), sc_cli::Error> {
+	let cli = TestCollatorCli::parse();
+
+	match &cli.subcommand {
+		Some(Commands::BuildSpec(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			runner.sync_run(|config| cmd.run(config.chain_spec, config.network))
+		},
+		Some(Commands::ExportGenesisState(params)) => {
+			let mut builder = sc_cli::LoggerBuilder::new("");
+			builder.with_profiling(sc_tracing::TracingReceiver::Log, "");
+			let _ = builder.init();
+
+			let parachain_id = ParaId::from(params.parachain_id);
+			let spec = Box::new(cumulus_test_service::get_chain_spec(parachain_id)) as Box<_>;
+			let state_version = cumulus_test_service::runtime::VERSION.state_version();
+
+			let block: parachains_common::Block = generate_genesis_block(&spec, state_version)?;
+			let raw_header = block.header().encode();
+			let output_buf = if params.raw {
+				raw_header
+			} else {
+				format!("0x{:?}", HexDisplay::from(&block.header().encode())).into_bytes()
+			};
+
+			if let Some(output) = &params.output {
+				std::fs::write(output, output_buf)?;
+			} else {
+				std::io::stdout().write_all(&output_buf)?;
+			}
+
+			Ok(())
+		},
+		Some(Commands::ExportGenesisWasm(params)) => {
+			let mut builder = sc_cli::LoggerBuilder::new("");
+			builder.with_profiling(sc_tracing::TracingReceiver::Log, "");
+			let _ = builder.init();
+
+			let parachain_id = ParaId::from(params.parachain_id);
+			let spec = Box::new(cumulus_test_service::get_chain_spec(parachain_id)) as Box<_>;
+			let raw_wasm_blob = extract_genesis_wasm(&spec)?;
+			let output_buf = if params.raw {
+				raw_wasm_blob
+			} else {
+				format!("0x{:?}", HexDisplay::from(&raw_wasm_blob)).into_bytes()
+			};
+
+			if let Some(output) = &params.output {
+				std::fs::write(output, output_buf)?;
+			} else {
+				std::io::stdout().write_all(&output_buf)?;
+			}
+
+			Ok(())
+		},
+		None => {
+			let mut builder = sc_cli::LoggerBuilder::new("");
+			builder.with_colors(true);
+			let _ = builder.init();
+
+			let collator_options = cli.run.collator_options();
+			let tokio_runtime = sc_cli::build_runtime()?;
+			let tokio_handle = tokio_runtime.handle();
+			let mut config = cli
+				.run
+				.normalize()
+				.create_configuration(&cli, tokio_handle.clone())
+				.expect("Should be able to generate config");
+
+			let parachain_id = ParaId::from(cli.parachain_id);
+			config.chain_spec =
+				Box::new(cumulus_test_service::get_chain_spec(parachain_id)) as Box<_>;
```

Comment:
Why do we need this? The `chain_spec` should already be the correct one through the `load_spec` above?